### PR TITLE
Expose custom dialog extension APIs

### PR DESCRIPTION
**Pull request by MattIrv**
This exposes the current version of the proposed custom dialog APIs for extension use. The APIs aren't finalized yet but this will make it easier to test and develop for now.

The following code can be used to register a command that shows a dialog when executed, assuming the `sqlservices` model view content provider is already registered:

```
	vscode.commands.registerCommand('mssql.openTestDialog', () => {
		let dialog = sqlops.window.modelviewdialog.createDialog('Test dialog');
		let tab1 = sqlops.window.modelviewdialog.createTab('Test tab 1');
		tab1.content = 'sqlservices';
		let tab2 = sqlops.window.modelviewdialog.createTab('Test tab 2');
		tab2.content = 'sqlservices';
		dialog.content = [tab1, tab2];
		dialog.okButton.onClick(() => console.log('ok clicked!'));
		dialog.cancelButton.onClick(() => console.log('cancel clicked!'));
		dialog.okButton.label = 'ok';
		dialog.cancelButton.label = 'no';
		let customButton1 = sqlops.window.modelviewdialog.createButton('Test button 1');
		customButton1.onClick(() => console.log('button 1 clicked!'));
		let customButton2 = sqlops.window.modelviewdialog.createButton('Test button 2');
		customButton2.onClick(() => console.log('button 2 clicked!'));
		dialog.customButtons = [customButton1, customButton2];
		dialog.open();
	});
```

Modal with custom content, custom buttons, and custom labels for the ok and cancel buttons

### CODE REVIEW

File: src/sql/workbench/api/common/sqlExtHostTypes.ts [hunk 1]
```diff
@@ -94,3 +94,21 @@ export interface IComponentEventArgs {
 	eventType: ComponentEventType;
 	args: any;
 }
+
+export interface IModelViewDialogDetails {
+	title: string;
+	content: string | number[];
+	okTitle: string;
+	cancelTitle: string;
```
Comment (kevcunnane) [inline src/sql/workbench/api/common/sqlExtHostTypes.ts]: Do we support the case where there's only an OK button? E.g. some cases where things will always succeed?

File: src/sql/workbench/api/node/extHostModelViewDialog.ts [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+import { IItemConfig, ModelComponentTypes, IComponentShape } from 'sql/workbench/api/common/sqlExtHostTypes';
+
+class DialogImpl implements sqlops.window.modelviewdialog.Dialog {
+	public title: string;
+	public content: string | TabImpl[];
+	public okTitle: string;
+	public cancelTitle: string;
+	public customButtons: ButtonImpl[];
+
+	private readonly _onOk = new Emitter<void>();
```
Comment (kevcunnane) [inline src/sql/workbench/api/node/extHostModelViewDialog.ts]: Minor: could we just wrap the OK/Cancel as Buttons to simplify this?

**Comment by MattIrv** [comment]
@kevcunnane I just updated the API to use buttons for ok and back like you suggested and implement the custom button handler. Can you take a look at the change and make sure it matches what you had in mind? https://github.com/Microsoft/sqlopsstudio/pull/1206/commits/f168485b71befb4b55d0316f780f1d4eb5ac5aac

I also updated the sample code above and included a screenshot showing the dialog created by that sample code, including 2 custom buttons and custom labels for the "Done" and "Cancel" buttons

File: src/sql/platform/dialog/dialogTypes.ts [hunk 3]
```diff
@@ -21,16 +22,14 @@ export class DialogTab implements sqlops.window.modelviewdialog.DialogTab {
 }
 
 export class Dialog implements sqlops.window.modelviewdialog.Dialog {
+	private static readonly DONE_BUTTON_LABEL = localize('dialogModalDoneButtonLabel', 'Done');
+	private static readonly CANCEL_BUTTON_LABEL = localize('dialogModalCancelButtonLabel', 'Cancel');
+
 	public content: string | DialogTab[];
-	public okTitle: string;
-	public cancelTitle: string;
+	public okButton: DialogButton = new DialogButton(Dialog.DONE_BUTTON_LABEL, true);
```
Comment (kevcunnane) [inline src/sql/platform/dialog/dialogTypes.ts]: Minor: sometimes people want to override the "Done" label. But this is fine for now

File: src/sql/workbench/api/node/extHostModelViewDialog.ts [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+class DialogImpl implements sqlops.window.modelviewdialog.Dialog {
+	public title: string;
+	public content: string | sqlops.window.modelviewdialog.DialogTab[];
+	public okButton: sqlops.window.modelviewdialog.Button;
+	public cancelButton: sqlops.window.modelviewdialog.Button;
+	public customButtons: sqlops.window.modelviewdialog.Button[];
+
+	constructor(private _extHostModelViewDialog: ExtHostModelViewDialog) {
+		this.okButton = this._extHostModelViewDialog.createButton(nls.localize('dialogOkLabel', 'Done'));
```
Comment (kevcunnane) [inline src/sql/workbench/api/node/extHostModelViewDialog.ts]: I'd suggest allowing users to update the ok button's label somewhere in this. Is this possible?

Comment (MattIrv) [reply src/sql/workbench/api/node/extHostModelViewDialog.ts]: The user can override the label, see my sample code for an example of this. Right now changes to the label while the dialog is already open won’t be reflected, but I plan to do future work to let extensions update all aspects of dialogs while they’re already open

**Pull request merged**